File: v2/lib/gallformers/explore.ex
defmodule Gallformers.Explore do
  @moduledoc """
  The Explore context.

  Provides functions for building hierarchical tree data for browsing
  galls and hosts by taxonomic family.
  """

  import Ecto.Query
  alias Gallformers.Repo
  alias Gallformers.Species.{Gall, GallSpecies, Species}
  alias Gallformers.Taxonomy.Taxonomy

  @doc """
  Returns a hierarchical tree of gall species organized by Family → Genus → Species.

  Each node has:
  - key: unique identifier (string)
  - label: display text
  - url: navigation URL (only for species leaves)
  - nodes: child nodes (for families and genera)
  """
  @spec get_galls_tree() :: [map()]
  def get_galls_tree do
    get_species_tree("gall", false)
  end

  @doc """
  Returns a hierarchical tree of undescribed gall species.
  """
  @spec get_undescribed_tree() :: [map()]
  def get_undescribed_tree do
    get_species_tree("gall", true)
  end

  @doc """
  Returns a hierarchical tree of host species organized by Family → Genus → Species.
  """
  @spec get_hosts_tree() :: [map()]
  def get_hosts_tree do
    get_species_tree("plant", false)
  end

  defp get_species_tree(taxoncode, undescribed_only) do
    rows = fetch_tree_data(taxoncode, undescribed_only)
    build_tree(rows, taxoncode)
  end

  defp fetch_tree_data("gall", undescribed_only) do
    base_query =
      from f in Taxonomy,
        join: g in Taxonomy,
        on: g.parent_id == f.id and g.type == "genus",
        join: st in "speciestaxonomy",
        on: st.taxonomy_id == g.id,
        join: s in Species,
        on: s.id == st.species_id,
        join: gs in GallSpecies,
        on: gs.species_id == s.id,
        join: gall in Gall,
        on: gs.gall_id == gall.id,
        where: f.type == "family" and f.description != "Plant" and s.taxoncode == "gall",
        order_by: [f.name, g.name, s.name],
        select: %{
          family_id: f.id,
          family_name: f.name,
          family_description: f.description,
          genus_id: g.id,
          genus_name: g.name,
          genus_description: g.description,
          species_id: s.id,
          species_name: s.name,
          undescribed: gall.undescribed
        }

    query =
      if undescribed_only do
        from [f, g, st, s, gs, gall] in base_query,
          where: gall.undescribed == true
      else
        base_query
      end

    Repo.all(query)
  end

  defp fetch_tree_data("plant", _undescribed_only) do
    from(f in Taxonomy,
      join: g in Taxonomy,
      on: g.parent_id == f.id and g.type == "genus",
      join: st in "speciestaxonomy",
      on: st.taxonomy_id == g.id,
      join: s in Species,
      on: s.id == st.species_id,
      where: f.type == "family" and f.description == "Plant" and s.taxoncode == "plant",
      order_by: [f.name, g.name, s.name],
      select: %{
        family_id: f.id,
        family_name: f.name,
        family_description: f.description,
        genus_id: g.id,
        genus_name: g.name,
        genus_description: g.description,
        species_id: s.id,
        species_name: s.name,
        undescribed: false
      }
    )
    |> Repo.all()
  end

  defp build_tree(rows, taxoncode) do
    url_prefix = if taxoncode == "plant", do: "/host/", else: "/gall/"

    # Group by family, then genus, then species
    # Using maps to track unique entries and maintain order
    families_map = %{}
    families_order = []

    {families_map, families_order} =
      Enum.reduce(rows, {families_map, families_order}, fn row, {fam_map, fam_order} ->
        family_key = row.family_id

        {fam_map, fam_order} =
          if Map.has_key?(fam_map, family_key) do
            {fam_map, fam_order}
          else
            family_entry = %{
              id: row.family_id,
              name: row.family_name,
              description: row.family_description,
              genera: %{},
              genera_order: []
            }

            {Map.put(fam_map, family_key, family_entry), fam_order ++ [family_key]}
          end

        # Update genera within family
        family = Map.get(fam_map, family_key)
        genus_key = row.genus_id

        {genera_map, genera_order} =
          if Map.has_key?(family.genera, genus_key) do
            {family.genera, family.genera_order}
          else
            genus_entry = %{
              id: row.genus_id,
              name: row.genus_name,
              description: row.genus_description,
              species: %{},
              species_order: []
            }

            {Map.put(family.genera, genus_key, genus_entry), family.genera_order ++ [genus_key]}
          end

        # Update species within genus
        genus = Map.get(genera_map, genus_key)
        species_key = row.species_id

        {species_map, species_order} =
          if Map.has_key?(genus.species, species_key) do
            {genus.species, genus.species_order}
          else
            species_entry = %{
              id: row.species_id,
              name: row.species_name
            }

            {Map.put(genus.species, species_key, species_entry),
             genus.species_order ++ [species_key]}
          end

        genus = %{genus | species: species_map, species_order: species_order}
        genera_map = Map.put(genera_map, genus_key, genus)
        family = %{family | genera: genera_map, genera_order: genera_order}
        fam_map = Map.put(fam_map, family_key, family)

        {fam_map, fam_order}
      end)

    # Convert to tree nodes
    Enum.map(families_order, fn family_key ->
      family = Map.get(families_map, family_key)
      genus_nodes = build_genus_nodes(family, url_prefix)

      %{
        key: "f-#{family.id}",
        label: format_label(family.name, family.description),
        nodes: genus_nodes
      }
    end)
  end

  defp build_genus_nodes(family, url_prefix) do
    Enum.map(family.genera_order, fn genus_key ->
      genus = Map.get(family.genera, genus_key)
      species_nodes = build_species_nodes(genus, url_prefix)

      %{
        key: "g-#{genus.id}",
        label: format_label(genus.name, genus.description),
        nodes: species_nodes
      }
    end)
  end

  defp build_species_nodes(genus, url_prefix) do
    Enum.map(genus.species_order, fn species_key ->
      species = Map.get(genus.species, species_key)

      %{
        key: "s-#{species.id}",
        label: species.name,
        url: "#{url_prefix}#{species.id}"
      }
    end)
  end

  defp format_label(name, nil), do: name
  defp format_label(name, ""), do: name
  defp format_label(name, "Plant"), do: name
  defp format_label(name, description), do: "#{name} (#{description})"
end

File: v2/api/internal/handlers/explore.go
package handlers

import (
	"database/sql"
	"log/slog"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
	db "github.com/jeffdc/gallformers/v2/api/internal/db/generated"
	"github.com/jeffdc/gallformers/v2/api/internal/middleware"
)

// ExploreHandler handles explore page HTTP requests.
type ExploreHandler struct {
	queries *db.Queries
}

// NewExploreHandler creates a new ExploreHandler.
func NewExploreHandler(queries *db.Queries) *ExploreHandler {
	return &ExploreHandler{queries: queries}
}

// TreeNode represents a node in the explore tree (family, genus, or species).
type TreeNode struct {
	Key   string     `json:"key"`
	Label string     `json:"label"`
	URL   string     `json:"url,omitempty"`
	Nodes []TreeNode `json:"nodes,omitempty"`
}

// ExploreResponse contains tree data for all three explore views.
type ExploreResponse struct {
	Galls       []TreeNode `json:"galls"`
	Undescribed []TreeNode `json:"undescribed"`
	Hosts       []TreeNode `json:"hosts"`
}

// RegisterRoutes registers explore routes on the router.
func (h *ExploreHandler) RegisterRoutes(r chi.Router) {
	r.Get("/explore", h.GetExploreData)
}

// GetExploreData handles GET /api/v2/explore
// Returns hierarchical tree data for galls, undescribed galls, and hosts.
func (h *ExploreHandler) GetExploreData(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	// Fetch all three datasets in parallel would be nice, but for simplicity fetch sequentially
	gallRows, err := h.queries.GetGallFamiliesWithSpecies(ctx)
	if err != nil {
		slog.Error("failed to get gall families", "error", err)
		middleware.RespondInternalError(w, "Failed to get explore data")
		return
	}

	undescribedRows, err := h.queries.GetUndescribedGallFamiliesWithSpecies(ctx)
	if err != nil {
		slog.Error("failed to get undescribed gall families", "error", err)
		middleware.RespondInternalError(w, "Failed to get explore data")
		return
	}

	hostRows, err := h.queries.GetHostFamiliesWithSpecies(ctx)
	if err != nil {
		slog.Error("failed to get host families", "error", err)
		middleware.RespondInternalError(w, "Failed to get explore data")
		return
	}

	response := ExploreResponse{
		Galls:       buildGallTree(gallRows),
		Undescribed: buildUndescribedTree(undescribedRows),
		Hosts:       buildHostTree(hostRows),
	}

	middleware.RespondOK(w, response)
}

// buildGallTree converts flat rows into a hierarchical tree structure for galls.
func buildGallTree(rows []db.GetGallFamiliesWithSpeciesRow) []TreeNode {
	return buildTree(rows, func(r db.GetGallFamiliesWithSpeciesRow) treeRow {
		return treeRow{
			FamilyID:          r.FamilyID,
			FamilyName:        r.FamilyName,
			FamilyDescription: r.FamilyDescription,
			GenusID:           r.GenusID,
			GenusName:         r.GenusName,
			GenusDescription:  r.GenusDescription,
			SpeciesID:         r.SpeciesID,
			SpeciesName:       r.SpeciesName,
			URLPath:           "gall",
		}
	})
}

// buildUndescribedTree converts flat rows into a hierarchical tree structure for undescribed galls.
func buildUndescribedTree(rows []db.GetUndescribedGallFamiliesWithSpeciesRow) []TreeNode {
	return buildTree(rows, func(r db.GetUndescribedGallFamiliesWithSpeciesRow) treeRow {
		return treeRow{
			FamilyID:          r.FamilyID,
			FamilyName:        r.FamilyName,
			FamilyDescription: r.FamilyDescription,
			GenusID:           r.GenusID,
			GenusName:         r.GenusName,
			GenusDescription:  r.GenusDescription,
			SpeciesID:         r.SpeciesID,
			SpeciesName:       r.SpeciesName,
			URLPath:           "gall",
		}
	})
}

// buildHostTree converts flat rows into a hierarchical tree structure for hosts.
func buildHostTree(rows []db.GetHostFamiliesWithSpeciesRow) []TreeNode {
	return buildTree(rows, func(r db.GetHostFamiliesWithSpeciesRow) treeRow {
		return treeRow{
			FamilyID:          r.FamilyID,
			FamilyName:        r.FamilyName,
			FamilyDescription: r.FamilyDescription,
			GenusID:           r.GenusID,
			GenusName:         r.GenusName,
			GenusDescription:  r.GenusDescription,
			SpeciesID:         r.SpeciesID,
			SpeciesName:       r.SpeciesName,
			URLPath:           "host",
		}
	})
}

// treeRow is a common structure for building trees.
type treeRow struct {
	FamilyID          int64
	FamilyName        string
	FamilyDescription sql.NullString
	GenusID           int64
	GenusName         string
	GenusDescription  sql.NullString
	SpeciesID         int64
	SpeciesName       string
	URLPath           string
}

// buildTree is a generic tree builder that works with any row type.
func buildTree[T any](rows []T, toRow func(T) treeRow) []TreeNode {
	// Build family -> genus -> species hierarchy
	// Use maps to track unique entries and maintain order
	type speciesEntry struct {
		id   int64
		name string
		url  string
	}

	type genusEntry struct {
		id          int64
		name        string
		description string
		species     []speciesEntry
		speciesMap  map[int64]bool
	}

	type familyEntry struct {
		id          int64
		name        string
		description string
		genera      []*genusEntry
		genusMap    map[int64]*genusEntry
	}

	families := make([]*familyEntry, 0)
	familyMap := make(map[int64]*familyEntry)

	for _, row := range rows {
		r := toRow(row)

		// Get or create family
		family, exists := familyMap[r.FamilyID]
		if !exists {
			family = &familyEntry{
				id:          r.FamilyID,
				name:        r.FamilyName,
				description: nullStringToString(r.FamilyDescription),
				genera:      make([]*genusEntry, 0),
				genusMap:    make(map[int64]*genusEntry),
			}
			familyMap[r.FamilyID] = family
			families = append(families, family)
		}

		// Get or create genus within family
		genus, exists := family.genusMap[r.GenusID]
		if !exists {
			genus = &genusEntry{
				id:          r.GenusID,
				name:        r.GenusName,
				description: nullStringToString(r.GenusDescription),
				species:     make([]speciesEntry, 0),
				speciesMap:  make(map[int64]bool),
			}
			family.genusMap[r.GenusID] = genus
			family.genera = append(family.genera, genus)
		}

		// Add species to genus (avoid duplicates)
		if !genus.speciesMap[r.SpeciesID] {
			genus.speciesMap[r.SpeciesID] = true
			genus.species = append(genus.species, speciesEntry{
				id:   r.SpeciesID,
				name: r.SpeciesName,
				url:  "/" + r.URLPath + "/" + intToString(r.SpeciesID),
			})
		}
	}

	// Convert to TreeNode structure
	result := make([]TreeNode, len(families))
	for i, family := range families {
		familyNode := TreeNode{
			Key:   intToString(family.id),
			Label: formatWithDescription(family.name, family.description),
			Nodes: make([]TreeNode, len(family.genera)),
		}

		for j, genus := range family.genera {
			genusNode := TreeNode{
				Key:   intToString(genus.id),
				Label: formatWithDescription(genus.name, genus.description),
				Nodes: make([]TreeNode, len(genus.species)),
			}

			for k, sp := range genus.species {
				genusNode.Nodes[k] = TreeNode{
					Key:   intToString(sp.id),
					Label: sp.name,
					URL:   sp.url,
				}
			}

			familyNode.Nodes[j] = genusNode
		}

		result[i] = familyNode
	}

	return result
}

// formatWithDescription formats a name with an optional description in parentheses.
func formatWithDescription(name, description string) string {
	if description == "" {
		return name
	}
	return name + " (" + description + ")"
}

// nullStringToString converts a sql.NullString to a string.
func nullStringToString(ns sql.NullString) string {
	if ns.Valid {
		return ns.String
	}
	return ""
}

// intToString converts an int64 to a string.
func intToString(n int64) string {
	return strconv.FormatInt(n, 10)
}
